src/csharp/Grpc.Core/ServerServiceDefinition.cs

```diff
@@ -45,6 +48,28 @@ private ServerServiceDefinition(Dictionary<string, IServerCallHandler> callHandl
             }
         }
 
+        /// <summary>
+        /// Returns a <see cref="Grpc.Core.ServerServiceDefinition" /> instance that
+        /// intercepts calls to the underlying service handler via the given interceptor.
+        /// This is an EXPERIMENTAL API.
+        /// </summary>
+        /// <param name="interceptor">The interceptor to register on service.</param>
+        public ServerServiceDefinition Intercept(Interceptor interceptor)
+        {
+            GrpcPreconditions.CheckNotNull(interceptor, "interceptor");
+            return new ServerServiceDefinition(CallHandlers.ToDictionary(
+                x => x.Key, x =>
+                {
+                    var value = x.Value;
+                    var interceptable = value as IInterceptableCallHandler;
```

Comment:
maybe I don't understand correctly, but it looks a bit like "manual" dispatch (in the sense of not utilizing polymorphism correctly). 

In what situation won't the value be not an instance of IInterceptableCallHandler?

Reply:
<strike>We could have fused the two interfaces together (i.e. `IInterceptableCallHandler` into `IServerCallHandler` and all of them are internal so it doesn't really make any difference. I think in particular I had the `Unimplemented` implementation in mind, but we can simply add a no-op method to that and merge the interfaces. </strike>

Merged it into the `IServerCallHandler` interface.